src/nvim/api/vim.c

```diff
@@ -164,6 +164,56 @@ Object vim_eval(String str, Error *err)
   return rv;
 }
 
+/// Call the given function with the given arguments stored in an array.
+///
+/// @param functionName The function to call
+/// @param args the functions arguments packed in an Array
+/// @param[out] err Details of an error that may have occurred
+/// @return The result of the function call if it was successful
+Object vim_call_function(String functionName, Array args, Error *err)
+  FUNC_ATTR_DEFERRED
+{
+  Object retv = OBJECT_INIT;
+  if (args.size > MAX_FUNC_ARGS) {
+    api_set_error(err, Exception,
+      _("Function called with too many arguments."));
+    return retv;
+  }
+
+  try_start();
+  // Convert the arguments in args from Object to typval_T values
+  typval_T vim_args[MAX_FUNC_ARGS + 1];
+  size_t i = 0;  // also used for freeing the variables
+  for (; i < args.size; i++) {
+    if (!object_to_vim(args.items[i], &vim_args[i], err)) {
+      api_set_error(err, Exception,
+          _("Error converting from object in vim_function."));
+      goto free_vim_args;
+    }
+  }
+
+  // Call the function
+  typval_T tmp;
+  int dummy;
+  int r = call_func((char_u*) functionName.data, (int)functionName.size,
+            &tmp, (int)args.size, vim_args,
+            curwin->w_cursor.lnum, curwin->w_cursor.lnum, &dummy, true, NULL);
```

Comment:
I mentioned a flaw: you have `NULL` for `self` dictionary here. Some user functions cannot possibly be called without a dictionary.